### PR TITLE
CRF layer v3.0 continued

### PR DESCRIPTION
**Pull request by jaspersjsun**
This is a continued PR of #1733 by @gabrieldemarmiesse .

Several suggestions from code review were applied.

Following is the original comment.

------

With a subclassing approch, we have a nicer API and it's very flexible.

Works only with TF 2.2+

@howl-anderson for the review and the CLA

The plan is to show users how to do the subclassing for the CRF. We shouldn't provide and API to save them some code there because it's going to become very complex to design a good API and to maintain it later on.

So the CRF layer is a public API and for the CRF loss, we give a good tutorial about subclassing.

Quick tutorial right now:

```Python

import tensorflow as tf
from tensorflow_addons.layers.crf import CRF
from tensorflow_addons.text.crf import crf_log_likelihood

def unpack_data(data):
    if len(data) == 2:
        return data[0], data[1], None
    elif len(data) == 3:
        return data
    else:
        raise TypeError("Expected data to be a tuple of size 2 or 3.")


class ModelWithCRFLoss(tf.keras.Model):
    """Wrapper around the base model for custom training logic."""

    def __init__(self, base_model):
        super().__init__()
        self.base_model = base_model

    def call(self, inputs):
        return self.base_model(inputs)

    def compute_loss(self, x, y, sample_weight, training=False):
        y_pred = self(x, training=training)
        _, potentials, sequence_length, chain_kernel = y_pred

        crf_loss = -crf_log_likelihood(potentials, y, sequence_length, chain_kernel)[0]

        if sample_weight is not None:
            crf_loss = crf_loss * sample_weight

        return tf.reduce_mean(crf_loss), sum(self.losses)

    def train_step(self, data):
        x, y, sample_weight = unpack_data(data)

        with tf.GradientTape() as tape:
            crf_loss, internal_losses = self.compute_loss(
                x, y, sample_weight, training=True
            )
            total_loss = crf_loss + internal_losses

        gradients = tape.gradient(total_loss, self.trainable_variables)
        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))

        return {"crf_loss": crf_loss, "internal_losses": internal_losses}

    def test_step(self, data):
        x, y, sample_weight = unpack_data(data)
        crf_loss, internal_losses = self.compute_loss(x, y, sample_weight)
        return {"crf_loss": crf_loss, "internal_losses": internal_losses}


x_np, y_np = get_test_data()

x_input = tf.keras.layers.Input(shape=x_np.shape[1:])
crf_outputs = CRF(5)(x_input)
base_model = tf.keras.Model(x_input, crf_outputs)
model = ModelWithCRFLoss(base_model)

model.compile("adam")
model.fit(x=x_np, y=y_np)
model.evaluate(x_np, y_np)
model.predict(x_np)
model.save("my_model.tf")

```

### CODE REVIEW

**Comment by FakeCola** [comment]
@googlebot I signed it!

**Comment by gabrieldemarmiesse** [comment]
@googlebot I conscent.

**Comment by howl-anderson** [comment]
@googlebot I signed it!

**Comment by FakeCola** [comment]
Wheel building processes failed on Ubuntu-18.04, Python3.5. Seems there are some problems with the docker env. Anything I need to do to pass those two checks?

**Comment by FakeCola** [comment]
@howl-anderson Thank you so much for your contribution. Would you mind replying with only `@googlebot I consent.` to grand consent to this PR?

**Comment by howl-anderson** [comment]
@googlebot I consent.

**Comment by FakeCola** [comment]
@facaiy @seanpmorgan Hi guys, would you mind having a look at this PR whenever you are available? The history of adding CRF layer can be dated back one year ago ( #22 #314 #377 and many other issues and PRs). Having this published will help a lot . Very appreciated!

**Comment by luozhouyang** [comment]
Thank you a lot for this contribution!

Can we use `tf.keras.layers.Layer.add_loss` API inside `CRF` layer to calculate the `crf_loss` instead of using a wrapper model like `ModelWithCRFLoss`?

**Comment by howl-anderson** [comment]
@luozhouyang The pattern that you said is called endpoint pattern I think. In that pattern, users need to pass true labels as one of the inputs, and I think it is not a user-friendly API. The tensorflow/tensorflow#37818 is a promising solution to this, but still under review.

**Comment by eloukas** [comment]
This approach can work but it's incredibly slow, especially when we have a vast amount of classes/labels.
Is there any way you can incorporate Viterbi decoding in it?
